### DISCUSSION TITLE
Can users/developers add to or edit their did:plc document?

### DISCUSSION THREAD

**Discussion by kgzoom** [bluesky-social/atproto#1892] [Protocol (atproto)]
Hi,
I'm assuming that a Bluesky server initializes a default DID document for the user that complies with basic requirements to be a did:plc, but developers (with user permission) could add custom data to the did:plc to extend its functionality? I'm not familiar with this topic, but I'm guessing that, once you get a default DID from your server that complies with the basic Bluesky requirements, the user has complete control of it, probably because that's required by the standard, and users/developers would be able to add information to the DID to extend its functionality, right? Also, where is the did:plc stored? always on your server, or does some DID service take care of storing it? Or is that up to the individual user?

**Comment by snarfed** [comment]
https://github.com/did-method-plc/did-method-plc#readme (and https://atproto.com/specs/did ) probably answer many of your questions. Short answers: `did:plc` is a centralized registry run by the Bluesky team. `did:web` is the more decentralized supported alternative. Users will eventually be able to modify their `did:plc` documents with their recovery keys. This is unusual though, and it's not how you add/change user-facing functionality in ATProto. That's done in clients, AppViews, and PDSes alone.

**Reply by kgzoom** [reply]
@snarfed So, one example, I'm guessing, of when they would use their recovery keys to edit their did:plc is when they change servers? What about adding user/developer "service endpoints". Is that not how the Bluesky team anticipates did:plc will be used? I assumed that DIDs would be a way of extending Bluesky's functionality? Or, does the Bluesky team prefer that only be done with did:web?

**Reply by snarfed** [reply]
Right, that and key rotation are the two main example use cases. I don't know of many others.

PDSes are designed to be application-, lexicon-, and data-independent, so it's not yet clear what you would use multiple service endpoints for, at least within ATProto, or how you'd distinguish them.

**Reply by snarfed** [reply]
...and again, extending functionality happens with new lexicons and AppViews, not DID doc changes.

**Reply by kgzoom** [reply]
Ok, so it would link to your server and PDS, and all the different types of functionality you want would be done there.

**Reply by snarfed** [reply]
In general, sure, and that's what ATProto does.

We're getting pretty abstract here, btw. Feel free to think about concrete use cases instead! Might help.

**Comment by bnewbold** [comment]
The short answer today is that philosophically accounts have control over their DID documents, but in practical terms they are controlled directly by the PDS. It is technically possible to create an account on the Bluesky PDS instance with a pre-existing PLC identity (including recovery key), but this is kind of buried away and only a handful of folks (out of millions of accounts) have done this.

We do intend to give accounts more control over their PLC identity via atproto API endpoints, including to configure recovery keys which can then be used to talk to the PLC directory directly. Those APIs still need to be designed and implemented, before federation ("identity control" in the roadmap: https://github.com/bluesky-social/atproto/discussions/1711).

Additional non-atproto services and identifiers ("also known as") could be added to DID documents in the future, PLC supports that. There will likely be a set of limits around DID documents: rate of updates (a few a day), overall size in bytes, etc. Inside those limits a lot of service and identity integrations should be possible.

**Reply by kgzoom** [reply]
@bnewbold Will the PDS instance be able to run inside a browser (using Wasm)?

**Comment by bnewbold** [comment]
To come back around to this, there are some new endpoints that were exposed a couple months ago as part of account migration which allow flexible control of PLC identities for existing accounts, including updating rotation keys to fully remove PDS host (eg, Bluesky) control over the identity.

This is all fairly low-level and not directly accessible in-app: you need to understand the PLC data structures to work with this. Details in this doc:
https://github.com/bluesky-social/pds/blob/main/ACCOUNT_MIGRATION.md

**Comment by filip26** [comment]
Hi,
 I understand that Bluesky main goal is not about DIDs, but it would nice to allow a user, or to an app on user's behalf and with consent, to manage it's own DID document, and use `did:plc` even for things that bsky team haven't thought of,  the design is flexible. At the end of the day, you could get 3rd party apps providing 'Authenticate with BSky', 'Sign a document with BSky ', etc. functionality almost for free.

If there is a better place where to share such ideas then please point me in the right direction.  Thank you.

**Reply by bnewbold** [reply]
We are totally in favor of this!
